frontend/src/registration/actions.ts

```diff
@@ -72,96 +76,124 @@ export const clearErrorMessage = (
-  data: RegistrationModel,
-  nextStep?: string
-): OcimThunkAction<void> => async dispatch => {
-  // Placeholder for form submit method
+  userData: RegistrationModel,
+  instanceData: RegistrationModel,
+  nextStep?: RegistrationSteps
+): OcimThunkAction<void> => async (dispatch: any) => {
   dispatch({
-    type: Types.REGISTRATION_SUBMIT,
-    data
+    type: Types.REGISTRATION_SUBMIT
   });
 
-  setTimeout(() => {
-    try {
-      // TODO
-      if (data.domain === 'existing') {
-        throw Error('Test error');
-      }
-      dispatch({ type: Types.REGISTRATION_SUCCESS, data });
-      if (nextStep) {
-        dispatch(push(nextStep));
-      }
-      dispatch({ type: UIActionTypes.NAVIGATE_NEXT_PAGE });
-    } catch (error) {
-      dispatch({
-        type: Types.REGISTRATION_FAILURE,
-        error
+  const userRegistrationData: any = { ...userData };
+  const registrationFeedback: any = {};
+
+  // Check for privacy policy and replace by current date
+  if (userRegistrationData.acceptTOS) {
+    delete userRegistrationData.acceptTOS;
+    userRegistrationData.acceptedPrivacyPolicy = new Date();
+  } else {
+    registrationFeedback.acceptTOS =
+      'You must accept the Terms of Service and Privacy policy.';
+  }
+
+  // Check if password confirmation is correct
+  if (userRegistrationData.password === userRegistrationData.passwordConfirm) {
+    delete userRegistrationData.passwordConfirm;
+  } else {
+    registrationFeedback.passwordConfirm =
+      'The password confirmation should match the password from the field above.';
+  }
+
+  if (Object.entries(registrationFeedback).length === 0) {
+    V2Api.accountsCreate({
+      data: { ...userRegistrationData }
+    })
+      .then(() => {
+        // Perform authentication and create new instance
+        dispatch(
+          performLogin({
+            username: userRegistrationData.username,
+            password: userRegistrationData.password
+          })
+        ).then(() => {
+          // Create instance
+          V2Api.instancesOpenedxConfigCreate({
+            data: { ...instanceData }
+          })
+            .then(() => {
+              dispatch({
+                type: Types.REGISTRATION_VALIDATION_SUCCESS,
+                userData,
+                nextStep
+              });
+              if (nextStep) {
+                dispatch(push(REGISTRATION_STEPS[nextStep]));
+              }
+            })
+            .catch(e => {
+              console.log("This isn't supposed to happen!", e);
+            });
+        });
+      })
+      .catch((e: any) => {
+        e.json().then((feedback: any) => {
+          // If validation fails, return error to form through state
+          const error = { ...feedback };
+          // Loop at each error message and join them.
+          // Also convert keys from snake_case to camelCase
+          Object.keys(error).forEach(key => {
+            error[toCamelCase(key)] = error[key].join();
+          });
```

Comment:
This code has appeared twice, and will likely show up again, so perhaps convert it into a utility function. Then this can become:

`const error = toCamelCaseKeys(feedback)`